gradle: pass through wrapper override arguments to unwrapped Gradle

numinit opened the pull request:
Comments on #436358 and #402297 indicated a desire to have a more controllable Gradle wrapper derivation. We were previously hiding the Gradle being wrapped in the closure of wrapGradle, forcing us to expose wrapGradle and use makeOverridable. Simply use callPackage to make it easier for consumers to override things.

@menixator @msfjarvis Does something like `pkgs.gradle.override { gradle-unwrapped = ... }` solve this for you?

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

numinit committed:
gradle: pass through wrapper override arguments to unwrapped Gradle

Comments on #436358 and #402297 indicated a desire to have a more
controllable Gradle wrapper derivation. We were previously hiding the
Gradle being wrapped in the closure of wrapGradle, forcing us to expose
wrapGradle and use makeOverridable. Simply use callPackage to make it
easier for consumers to override things without us needing to expose gradleGen.

numinit commented:
This shouldn't break anything, so can be included in 25.11 too.

msfjarvis commented:
How do I build an unwrapped Gradle out of tree? My main problem was not being able to build from pre-release versions without resorting to overrideAttrs on an existing derivation. That was why I ended up exposing the `gen'` method in my fork.

numinit committed:
gradle: expose gradle-packages.mkGradle and gradle-packages.wrapGradle

Users may want to use a completely custom version of Gradle. Add
mkGradle and wrapGradle as non-toplevel APIs for this.

numinit committed:
aliases: add gradleGen and wrapGradle

numinit commented on pkgs/top-level/aliases.nix:
```diff
@@ -675,6 +675,7 @@ mapAliases {
   graalvm-oracle = throw "'graalvm-oracle' has been renamed to/replaced by 'graalvmPackages.graalvm-oracle'"; # Converted to throw 2025-10-27
   graalvmCEPackages = throw "'graalvmCEPackages' has been renamed to/replaced by 'graalvmPackages'"; # Converted to throw 2025-10-27
   gradience = throw "`gradience` has been removed because it was archived upstream."; # Added 2025-09-20
+  gradleGen = throw "'gradleGen' has been moved to `gradle-packages.mkGradle`."; # Added 2025-11-02
```
Technically these were removed so it is nice to indicate it to users.

numinit commented on pkgs/development/tools/build-managers/gradle/default.nix:
```diff
@@ -16,81 +16,79 @@ let
       runCommand,
-    }:
-    this-gradle-unwrapped:
-    lib.makeOverridable (
```
Technically when you callPackage something, it becomes overridable. So we didn't need this function returning a function returning an overridable, we can just return the top function.

wolfgangwalther reviewed (approved):
Just glanced over it, didn't see any obvious problems.
